snapcraft/internal/build_providers/_lxd/_lxd.py

```diff
@@ -45,30 +45,49 @@ class LXD(Provider):
 
     @classmethod
     def ensure_provider(cls):
-        try:
-            if repo.snaps.SnapPackage.is_snap_installed("lxd"):
-                return
-        except repo.errors.SnapdConnectionError as snapd_error:
-            if sys.platform == "linux":
+        error_message = None  # type: Optional[str]
+        prompt_installable = False
+
+        if sys.platform != "linux":
+            error_message = "LXD is not supported on this platform"
+        else:
+            try:
+                if not repo.snaps.SnapPackage.is_snap_installed("lxd"):
+                    error_message = (
+                        "The LXD snap is required to continue: snap install lxd."
+                    )
+                    prompt_installable = True
+            except repo.errors.SnapdConnectionError:
                 error_message = (
                     "snap support is required to continue: "
                     "https://docs.snapcraft.io/installing-snapd/6735"
                 )
-            else:
-                error_message = "LXD is not supported on this platform"
+
+        if error_message is not None:
             raise errors.ProviderNotFound(
                 provider=cls._get_provider_name(),
-                prompt_installable=False,
+                prompt_installable=prompt_installable,
                 error_message=error_message,
-            ) from snapd_error
-
-        # At this point, we know that if we are on linux that we have
-        # snap support.
-        raise errors.ProviderNotFound(
-            provider=cls._get_provider_name(),
-            prompt_installable=True,
-            error_message="The LXD snap is required to continue: snap install lxd.",
-        )
+            )
+
+        # If we reach this point, it means the lxd snap is properly setup.
+        # Now is the time for additional sanity checks to ensure the provider
+        # will work.
+        try:
+            # TODO: add support for more distributions. Maybe refactor a bit so that Repo behaves
+            # similar to a build provider.
+            if repo.Repo.is_package_installed("lxd") or repo.Repo.is_package_installed(
+                "lxd-client"
+            ):
+                raise SnapcraftEnvironmentError(
```

Comment:
A comment on what happened to me when I allowed snapcraft to install the lxd snap (I was previously using the deb packages): it installed lxd, and successfully built my package. Then in the next attempt to run snapcraft, it refused to run because I had the deb packages installed -- but it ran the first time. Also I lost my custom default profile in the process and had to recreate it in the new lxd installation -- I wonder what else from my previous configuration has changed.

Reply:
The first time, before doing the iterative PRs to make this smaller involved setting LXD_DIR, then with a refactor I removed it as it was mostly a hack. So a mixup of deb and snap went into place.

pylxd does not support both installations at the same time in an official capacity but can be made to work.

I want to work the lxd team to get the best mechanisms to have lxd tell me where its socket lived, alas it was determined that we needed to hardcode the path to it (seen in later versions of this PR).

We do nothing special with the profile. But I do suspect having both installed at a point in time triggered `lxd init --auto` from /usr/bin instead of /snap/bin and that is now solved with the abspath to the binary and that process might have wiped your profile.

To solve the deb problem, we now just decide to falloff if the deb is installed.

Lastly, I want to add that to not lose stuff you need to run lxd.migrate from the snap to move away from the deb.

Reply:
lxd.migrate as part of commit 13cf4b9b